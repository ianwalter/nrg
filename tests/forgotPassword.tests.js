const { test } = require('@ianwalter/bff')
const app = require('../examples/accounts')
const { accounts } = require('../seeds/01_accounts')
const { extractEmailToken, getTestEmail } = require('..')

const generalUser = accounts.find(a => a.firstName === 'General')
const disabledUser = accounts.find(a => a.firstName === 'Disabled')

test('Forgot Password with invalid emails', async t => {
  let response = await app.test('/forgot-password').post({ email: null })
  t.expect(response.status).toBe(400)
  t.expect(response.body).toMatchSnapshot()

  response = await app.test('/forgot-password').post({ email: '' })
  t.expect(response.status).toBe(400)
  t.expect(response.body).toMatchSnapshot()

  const email = 'babu_frik@example'
  response = await app.test('/forgot-password').post({ email })
  t.expect(response.status).toBe(400)
  t.expect(response.body).toMatchSnapshot()
})

test('Forgot Password with unregistered email', async t => {
  const email = 'babu_frik@example.com'
  const response = await app.test('/forgot-password').post({ email })
  t.expect(response.status).toBe(200)
  t.expect(response.body).toMatchSnapshot()
})

test('Forgot Password with registered email', async t => {
  const response = await app.test('/forgot-password').post(generalUser)
  t.expect(response.status).toBe(200)
  t.expect(response.body).toMatchSnapshot()

  await t.asleep(500)

  // Extract and verify the Forgot Password email and token.
  const byEmail = email => email.headers.to === generalUser.email
  const { email } = await extractEmailToken(byEmail)
  t.expect(email).toMatchSnapshot({
    id: t.expect.any(String),
    messageId: t.expect.any(String),
    source: t.expect.any(String),
    date: t.expect.any(String),
    time: t.expect.any(String),
    envelope: {
      remoteAddress: t.expect.any(String)
    },
    headers: t.expect.any(Object)
  })

  // TODO: verify token database record.
})

test('Forgot Password for disabled user', async t => {
  const response = await app.test('/forgot-password').post(disabledUser)
  t.expect(response.status).toBe(200)
  t.expect(response.body).toMatchSnapshot()

  // Verify no email was sent to the user.
  const email = await getTestEmail(e => e.headers.to === disabledUser.email)
  t.expect(email).toBe(undefined)
})
