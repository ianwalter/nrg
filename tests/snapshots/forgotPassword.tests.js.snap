// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forgot password email required validation 1`] = `
Object {
  "feedback": Object {
    "email": Array [
      "A valid email is required.",
    ],
    "password": Array [
      "A valid password is required.",
    ],
  },
  "message": "Validation Error",
}
`;
